### ISSUE TITLE
Android Download Model Not working

### ISSUE THREAD

**Issue by zainaliyt** [ultralytics/hub#910]
### Search before asking

- [X] I have searched the HUB [issues](https://github.com/ultralytics/hub/issues) and found no similar bug report.

### HUB Component

Models

### Bug

When trying to download on android 14, the button does nothing but on ios it works as it downloads and runs my custom model.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @zainaliyt, thank you for raising an issue about [Ultralytics HUB](https://github.com/ultralytics/hub) 🚀! Please check out our [HUB Docs](https://docs.ultralytics.com/hub) to uncover more insights:

- [**Quickstart**](https://docs.ultralytics.com/hub/quickstart). Begin your journey by training and deploying YOLO models with HUB swiftly.
- [**Datasets: Preparing and Uploading**](https://docs.ultralytics.com/hub/datasets). Discover how to format and upload your datasets to HUB in YOLO layout.
- [**Models: Training and Exporting**](https://docs.ultralytics.com/hub/models). Train and export your YOLOv5 and YOLOv8 models on custom datasets.
- [**Ultralytics HUB App**](https://docs.ultralytics.com/hub/app). Uncover the potential of the Ultralytics App on iOS and Android, enabling direct model execution on mobile gadgets.
    * [**Android**](https://docs.ultralytics.com/hub/app/android). Explore TFLite model acceleration on Android devices.
    * [**iOS**](https://docs.ultralytics.com/hub/app/ios). Learn about optimizations for Apple's Neural Engine.

If this is a 🐛 Bug Report, could you please share more details, including screenshots and a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) on your Android setup? This will assist us in digging deeper into the issue you're facing with downloading models on Android 14 📱.

We aim to address all issues swiftly. An Ultralytics engineer will be here shortly to help further. Thank you for your patience and collaboration! 🌟

**Comment by zainaliyt** [comment]
I checked logs and prints from the app, when i press on the download app its prints downloading: null. Maybe this helps.

**Comment by pderrenger** [comment]
Hi @zainaliyt,

Thank you for providing additional details about the issue. It seems like there might be a problem with the model download functionality on Android 14. Here are a few steps you can take to troubleshoot and potentially resolve the issue:

1. **Update the App**: Ensure that you are using the latest version of the Ultralytics Android App. Updates often include bug fixes and improvements.

2. **Check Permissions**: Verify that the app has the necessary permissions to access storage on your device. Sometimes, permission issues can prevent downloads from completing successfully.

3. **Clear Cache**: Try clearing the app's cache and data. This can sometimes resolve unexpected behavior.

4. **Reinstall the App**: If the issue persists, consider uninstalling and reinstalling the app to ensure a fresh start.

5. **Test on Another Device**: If possible, try downloading the model on a different Android device to see if the issue is specific to your current device.

If none of these steps resolve the issue, it might be a bug specific to Android 14. Please ensure that your app and device are up-to-date, and let us know if the problem persists. We appreciate your patience and collaboration in helping us improve the app! 😊

If you have any further questions or need additional assistance, feel free to ask. We're here to help!

**Comment by yogendrasinghx** [comment]
Thank you for reporting the issue, @zainaliyt. Our team will check into this and begin working on a fix soon. We’ll keep you updated on the progress.

It would be very helpful if you could share your model ID (you can find this in the URL when using HUB Web) and details about your Android device hardware, including the model number. This information will assist us in diagnosing the problem more effectively.

**Comment by mymoomin** [comment]
I'm having the same issue, where clicking the download button doesn't do anything. It has every permission and I've tried reinstalling the app. I have version 0.9.25 of the app. The model ID is f7Rjtn7thHpxEnat4aJV. I have a Pixel 6a (model number G1AZG) running Android 14

**Comment by zainaliyt** [comment]
It somehow only works on ios devices.

Den sön 17 nov. 2024 15:23mymoomin ***@***.***> skrev:

**Comment by pderrenger** [comment]
Hi @yAlqubati,

Thank you for confirming you're experiencing the same issue and providing your Android version. It's helpful to know that this is affecting users on Android 13 as well, which broadens the scope of devices potentially impacted.

Here’s what we know so far:
- The issue occurs when pressing the download button on the Ultralytics Android App, which doesn't successfully download models.
- It seems to affect multiple Android versions (13 and 14) and devices (Pixel 6a in prior reports).
- The issue doesn't occur on iOS devices.

### Next Steps:
1. **App Update**: Ensure you are using the latest version of the Ultralytics App available on the [Google Play Store](https://play.google.com/store/apps/details?id=com.ultralytics.ultralytics_app).
2. **Check Network Connectivity**: The download process might fail if there's temporary network or server hiccups. Confirm your internet connection is stable and retry.
3. **Reproducibility Check**: Does the issue occur with all models in your account or only specific ones? If it’s model-specific, please share the model ID.

### Debugging Assistance:
If you can, please check if the app logs (e.g., via Android's `logcat`) display any specific errors when you press the download button. Additional technical details will help our team locate the root cause.

### Temporary Alternatives:
As a workaround, you could attempt using the browser-based Ultralytics HUB platform ([hub.ultralytics.com](https://hub.ultralytics.com)) to export your trained model in a preferred format like TensorFlow Lite (.tflite) or ONNX and then manually use it in an Android application.

We appreciate your patience as we dig deeper into this issue. I've logged this case for the development team to investigate further, and we'll keep you updated as we work toward a resolution!

If you have any other information or questions, feel free to share them here! 😊

**Comment by zainaliyt** [comment]
How do you change delegate?

**Comment by pderrenger** [comment]
Hi @zainaliyt,

Thanks for your question! To change the delegate for model inference in the Ultralytics Android App, follow these steps:

1. **Launch the App**: Open the Ultralytics Android App on your device.
2. **Navigate to Settings**: Go to the app's settings where you should see options for configuring your model's runtime.
3. **Select a Delegate**: Look for the delegate selection option. Depending on your device, you should see options like:
   - **CPU**: Default option for all devices, more consistent but slower.
   - **GPU**: Utilizes your device's GPU for faster inference.
   - **NNAPI**: The Android Neural Networks API, which can offload inference tasks to accelerators.
   - **Hexagon DSP**: Available on Qualcomm Snapdragon devices for efficient processing.
4. **Apply and Save**: After selecting a delegate, apply the changes and proceed with inference.

### Delegate Support Notes:
- Delegate availability depends on the hardware capabilities of your Android device. For instance, you need a Snapdragon processor for the Hexagon DSP delegate and a capable GPU for the GPU delegate.
- Test different delegates to identify the one that offers the best performance for your specific device and model [reference details here](https://docs.ultralytics.com/hub/app/android/#delegates-and-performance-variability).

Let us know if you need further assistance or have other questions. Happy experimenting! 😊

**Comment by zainaliyt** [comment]
Are you sure there is settings option in the android app? ![Screenshot_20241203_053938.jpg](https://github.com/user-attachments/assets/6636a145-aa1f-466f-9f3c-747b1f8ab12f)

**Comment by pderrenger** [comment]
Hi @zainaliyt,

Thank you for pointing that out! Currently, the Ultralytics Android App does not have a "Settings" menu where you can directly choose the inference delegate (e.g., CPU, GPU, NNAPI) through the UI. I apologize for any confusion in my earlier explanation. Delegate assignment is generally handled automatically by TensorFlow Lite in the app to prioritize performance based on the device's hardware.

If you're looking to manually specify a delegate (e.g., GPU instead of CPU), this feature isn't exposed via the app's UI. However, developers integrating YOLO models into their own Android apps can set specific delegates programmatically in their TensorFlow Lite interpreter configuration.

### Example: Specifying a Delegate in Code
For developers:
```java
Interpreter.Options options = new Interpreter.Options();
GpuDelegate gpuDelegate = new GpuDelegate();
options.addDelegate(gpuDelegate);

Interpreter tflite = new Interpreter(modelFile, options);
```

In the app, the process is likely automated by selecting the most performant delegate available based on your device hardware.

### Recommendation:
If you’re looking to run your model with a specific delegate (e.g., GPU or NNAPI), and this functionality is crucial for your use case, I recommend:
1. Export your model in TFLite format using [Ultralytics HUB](https://hub.ultralytics.com).
2. Integrate it into a custom Android app where you can programmatically set the delegate.

Our team is always working to improve the app's functionality, and your feedback is very valuable for shaping future features. If delegate selection is something you'd like to see in the app's settings, feel free to suggest it as a [feature request](https://github.com/ultralytics/hub/issues/new?assignees=&labels=enhancement&template=feature_request.md&title=%5BFeature%5D). 😊

Thanks for pointing this out, and let me know if you have any other questions or require further clarification!

**Comment by zainaliyt** [comment]
I am currently trying to use my yolo model in the ultralytics hub android app but cannot download the model in the app as it does not do anything, can you maybe fix and release a new version.

**Comment by john-rocky** [comment]
@zainaliyt 
Sorry for the inconvenience.
We have released version 9.31 of hub-app on the play store, which fixes the issue of not being able to download custom models for Android. We apologize for the inconvenience, but please install this and try again.
